### PR TITLE
Graph Importer configuration and improvements

### PR DESCRIPTION
**Pull request by manicminer**
Improvements to the Graph Importer

* Configurable import of services and versions similar to Resource Manager
* Re-enabling the Graph Importer workflow
* Support for `text/plain` responses
* Consolidate integer and float types to always use `int64` and `float64` respectively (maintainer-friendly)
* Explicit double singularization/pluralization in order to homogenize resource names
* Hardcode a workaround for naming inconsistency for `CloudPC` models and methods
* Resource ID and Resource naming changes to accommodate complex resources (see below)
* More code comments throughout now that the implementation is stabilizing
* Additional `list-tags` command to list available service tags in human-friendly table, to aid in future expansion

This change includes the following mitigations to prevent clobbering of resources and operations. Unfortunately, whilst it is nice to have shortened resource names to reduce boilerplate and improve readability, it isn't possible due to various services having complex URIs and/or very similar but semantically different URIs.

1. All resource names now include the slug `ById` whenever a `SegmentUserValue` is encountered in the URI
   e.g. `GetApplicationById` vs. `GetApplication`

2. All operations now use the long form resource name
   e.g. `ListApplicationByIdSynchronizationJobs` vs. `ListSynchronizationJobs`

3. OData suffixes are no longer trimmed from operation names
   e.g. `ListApplicationOwnerRefs` vs. `ListApplicationOwners`

4. The final segment is not singularized when constructing the resource name, so that action verbs can be reliably matched when parsing operations

5. Resource names are no longer de-duplicated, although this has minimal impact now that long-form names are used everywhere

In addition to operation names, these changes extend to resource IDs. Examples of the updated naming include: `ApplicationById`, `ApplicationByIdSynchronizationJobById`, `GroupByIdSiteByIdTermStoreGroupByIdSetByIdChildrenByIdChildrenByIdRelationById`.

Screenshot 2023-08-08 at 11 17 31

This enables us to support resources that can be addressed in complex ways that infer relationships, without needing complex name mangling rules which are likely to fall apart in the face of future additions to the API. In the v1.0 API alone, there are around 1000 resource conflicts due to this current ambiguity. Additionally, this also makes way for alternative IDs where resources are addressed by unique keys other than their primary resource/object ID, for example `/applications(appId='{appId}')`

Examples of complex/ambiguous URIs include:
```
/applications/{applicationId}/synchronization/jobs/validateCredentials
/applications/{applicationId}/synchronization/jobs/{jobId}/validateCredentials
```
```
/agreements/{agreementId}/file  (not a List operation)
/agreements/{agreementId}/files
/agreements/{agreementId}/files/{fileId}
```
```
/deviceAppManagement/mobileAppCategories
/deviceAppManagement/mobileApps/categories
/deviceAppManagement/mobileApps/categories/{categoryId}
```
```
/me/calendar  (not a List operation)
/me/calendar/{calendarId}
```
```
/users/{userId}/calendarView/{eventId}
/users/{userId}/calendars/{calendarId}/calendarView/{eventId}
/users/{userId}/calendar/events/{eventId}/instances/{eventId1}
/users/{userId}/calendars/{calendarId}/events/{eventId}/instances/{eventId1}
/users/{userId}/onenote/notebooks/{notebookId}/sectionGroups/{sectionGroupId}
/users/{userId}/onenote/notebooks/{notebookId}/sectionGroups/{sectionGroupId}/sectionGroups/{sectionGroupId1}
```
```
/sites/{siteId}/termStores/{storeId}/sets/{setId}/parentGroup/sets/{setId1}/children/{termId}/relations/{relationId}
/sites/{siteId}/termStore/sets/{setId}/parentGroup/sets/{setId1}/children/{termId}/children/{termId1}/relations/{relationId}
```
```
/groups/{groupId}/calendarView/{eventId}
/groups/{groupId}/calendar/calendarView/{eventId}
/groups/{groupId}/calendarView/{eventId}/instances/{eventId1}
/groups/{groupId}/calendar/calendarView/{eventId}/instances/{eventId1}
/groups/{groupId}/calendarView/{eventId}/instances/{eventId1}/extensions/{extensionId}
/groups/{groupId}/calendar/calendarView/{eventId}/instances/{eventId1}/extensions/{extensionId}
/groups/{groupId}/sites/{siteId}/termStore/groups/{groupId1}/sets/{setId}/children/{termId}/children/{termId1}
/groups/{groupId}/sites/{siteId}/termStores/{storeId}/groups/{groupId1}/sets/{setId}/children/{termId}/children/{termId1}
/groups/{groupId}/sites/{siteId}/termStores/{storeId}/groups/{groupId1}/sets/{setId}/children/{termId}/relations/{relationId}
/groups/{groupId}/sites/{siteId}/termStores/{storeId}/groups/{groupId1}/sets/{setId}/children/{termId}/children/{termId1}/relations/{relationId}
```

Pending review, this PR also re-enables the msgraph-metadata-importer workflow as all conflicts have been resolved in the currently-imported services:

Screenshot 2023-08-08 at 03 38 57

Screenshot 2023-08-08 at 03 36 15

### CODE REVIEW

**Comment by manicminer** [comment]
Unfortunately this approach is starting to fall apart for Identity Governance. URIs like this require more thought:

`/identityGovernance/entitlementManagement/accessPackageAssignments/{accessPackageAssignmentId}/accessPackage/accessPackageResourceRoleScopes/{accessPackageResourceRoleScopeId}/accessPackageResourceRole/accessPackageResource/accessPackageResourceScopes/{accessPackageResourceScopeId}/accessPackageResource/accessPackageResourceEnvironment`

**Comment by manicminer** [comment]
Added configuration to the importer to import services selectively, and excluding Identity Governance for now.

File: tools/importer-msgraph-metadata/pipeline/parser_models.go [hunk 1]
```diff
@@ -294,6 +271,13 @@ func parseCommonModels(schemas openapi3.Schemas) (models Models, err error) {
 		}
 	}
 
+	// Several constants are presented as models, to clean these up we'll remove any models having no fields
```
Comment (tombuildsstuff) [inline tools/importer-msgraph-metadata/pipeline/parser_models.go:274]: given they have no values, should we also be converting any references to string/etc fields?

File: tools/importer-msgraph-metadata/pipeline/task_parse_resources.go [hunk 2]
```diff
@@ -130,26 +123,38 @@ func (p pipelineTask) parseResourcesForService(resourceIds ResourceIds, models M
...
 									if f.Format == "binary" {
 										responseType = pointerTo(DataTypeBinary)
 										break
 									}
 
+									// Derive model and response type from title
 									if title := f.Title; title != "" || f.Type != "" {
-										if strings.HasPrefix(strings.ToLower(title), "collection of ") {
-											title = title[14:]
+										if strings.HasPrefix(title, "CollectionOf") {
```
Comment (tombuildsstuff) [inline tools/importer-msgraph-metadata/pipeline/task_parse_resources.go:152]: pedantic, but technically should this be insensitive given the input source?

**Review by tombuildsstuff — APPROVED** [review]
Left a few comments inline but this otherwise LGTM 👍

File: tools/importer-msgraph-metadata/pipeline/task_parse_resources.go (hunk 2 above)
Comment (manicminer) [reply tools/importer-msgraph-metadata/pipeline/task_parse_resources.go:152]: good shout 👍

File: tools/importer-msgraph-metadata/pipeline/task_parse_resources.go [hunk 3]
```diff
@@ -306,6 +311,12 @@ func (p pipelineTask) parseResourcesForService(resourceIds ResourceIds, models M
 				}
 			}
 		}
+
+		for _, operation := range resource.Operations {
+			if len(operation.Tags) > 1 {
+				fmt.Printf("%#v\n", operation.Tags)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (manicminer) [reply tools/importer-msgraph-metadata/pipeline/task_parse_resources.go:319]: I'll remove this, it was indeed just for debugging

File: tools/importer-msgraph-metadata/pipeline/parser_models.go (hunk 1 above)
Comment (manicminer) [reply tools/importer-msgraph-metadata/pipeline/parser_models.go:274]: At the moment there don't appear to be any direct references to these counterfeit models, I've added additional checking to make sure (and also to ensure that model references in general do exist). These will fall back to an underlying type, additionally reverting to string where the type cannot be determined at all.

File: tools/importer-msgraph-metadata/pipeline/normalize.go [hunk 4]
```diff
@@ -76,6 +76,9 @@ func cleanName(name string) string {
 	name = regexp.MustCompile("[^a-zA-Z0-9]").ReplaceAllString(name, "")
 	name = regexp.MustCompile("^Odata").ReplaceAllString(name, "OData")
 	name = regexp.MustCompile("^Innererror").ReplaceAllString(name, "InnerError")
+
+	// known issue where CloudPC appears with inconsistent casing
```
[the remark this answers is not in the document]
Comment (manicminer) [reply tools/importer-msgraph-metadata/pipeline/normalize.go:80]: Will raise one on the microsoftgraph/msgraph-metadata repo

**Pull request merged**